PySRSequenceRegressor

wenbang24 opened the pull request:
Resolve #94 

Adds ability to use symbolic regression on 1D recurrent array.
Added new keyword argument to PySRRegressor called `recursive_history_length`. Must be at least 1, or will throw a `ValueError`.

Here an example of how it works:
Suppose you have the array `[1,2,3,4,5,6,7,8,9]` and `recursive_history_length = 3`.
Then the feature array will be
```
[[1 2 3]
 [2 3 4]
 [3 4 5]
 [4 5 6]
 [5 6 7]
 [6 7 8]]
```
And the target array will be:
```
[4, 5, 6, 7, 8, 9]
```

And it's the same from there.

wenbang24 committed:
Added recursive_history_length

honestly I think it works
it doesn't take an target array - only features :)

wenbang24 committed:
Changed minimum recursive history length

from 1 to 0

wenbang24 committed:
fixed syntax error

wenbang24 committed:
Added recurrence functionality

wenbang24 committed:
Removed a debug print, also formatted

MilesCranmer commented:
Thanks! Regarding the design, I wonder if this shouldn’t go into PySRRegressor itself, but rather be a separate input transform that you would compose with scikit-learn’s Pipeline: https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html#sklearn.pipeline.Pipeline

This would be nice because it would automatically update `.predict` as well without extra internal code.

Basically I think PySRRegressor is already too large and I’d like to break it into smaller, composable pieces, so it’s easier for users to implement custom behavior.

Also please add some tests, as well as tests for how this interacts with `X_units`, `variable_names`, `weights`, etc.

wenbang24 committed:
new PySRSequenceRegressor class!

it inherits PySRRegressor and changes __init__ (new recursive_history_length hyperparameter) and run (preprocessing data so it works with everything else)
also got rid of stuff in PySRRegressor
also changed __init__.py to import new class

wenbang24 committed:
made recursive_history_length not optional

wenbang24 committed:
added tests for PySRSequenceRegressor

a lot copied from the PySRRegressor tests :)

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wenbang24 commented:
test_sequence_weighted_bumper test still isn't passing; how do the weights work?

MilesCranmer commented on pysr/sr.py:
```diff
@@ -2586,3 +2583,421 @@ def _mutate_parameter(param_name: str, param_value):
+        Parameters
+        ----------
+        X : ndarray | pandas.DataFrame
+            Training data of shape (n_samples, n_features).
```
docstring should be updated to time series data

wenbang24 committed:
changed __init__ of PySRSequenceRegressor to use PySRRegressor's __init__ function

wenbang24 committed:
fixed bug that removed first data point

wenbang24 committed:
added sequence to test names to make things a bit clearer

wenbang24 committed:
added .eggs to .gitignore

wenbang24 committed:
Merge branch 'recurrence' into wenbang24/issue94

wenbang24 committed:
Merge pull request #1 from wenbang24/wenbang24/issue94

oops wrong branch

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wenbang24 commented:
One test (test_sequence_noisy_builtin_variable_names) isn't passing because set_params doesn't work when PySRSequenceRegressor calls super().__init__()

wenbang24 committed:
updated docstring for PySRSequenceRegressor

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

MilesCranmer commented on pysr/sr.py:
```diff
@@ -2586,3 +2583,190 @@ def _mutate_parameter(param_name: str, param_value):
+        X = np.array(X)
+        y = y[self.recursive_history_length :]
+        y_units = X_units
+
```
Would it be simpler if everything after this point be done with `super().fit(X, y)`?

MilesCranmer commented on pysr/sr.py:
```diff
@@ -2586,3 +2583,190 @@ def _mutate_parameter(param_name: str, param_value):
+                f"Recursive symbolic regression with a history length of {self.recursive_history_length} requires at least {self.recursive_history_length + 2} datapoints."
+            )
+        y = X.copy()
+        X = []
+        for i in range(self.recursive_history_length, len(y)):
+            X.append(y[i - self.recursive_history_length : i].flatten())
+        X = np.array(X)
```
This should also be used in a `def predict(self, ...):` method to construct the input data for `super().predict(`

MilesCranmer commented on pysr/sr.py:
```diff
@@ -2586,3 +2583,190 @@ def _mutate_parameter(param_name: str, param_value):
+            for that particular element of `X`. Alternatively,
+            if a custom `loss` was set, it will can be used
+            in arbitrary ways.
+        variable_names : list[str]
+            A list of names for the variables, rather than "x0", "x1", etc.
+            If `X` is a pandas dataframe, the column names will be used
+            instead of `variable_names`. Cannot contain spaces or special
+            characters. Avoid variable names which are also
+            function names in `sympy`, such as "N".
+            The number of variable names must be equal to recurrence_history_length.
```
I think these should be automatically generated and include information about relative time. For example, `X_{t-1}`, `X_{t-2}` might be good choices?

MilesCranmer commented on pysr/sr.py:
```diff
@@ -2586,3 +2583,190 @@ def _mutate_parameter(param_name: str, param_value):
+        ----------
+        X : ndarray | pandas.DataFrame
+            Training time series data of shape (n_samples, 1) or (1, n_samples).
+        Xresampled : ndarray | pandas.DataFrame
+            Resampled training data, of shape (n_resampled, 1) or (1, n_resampled),
+            to generate a denoised data on. This
+            will be used as the training data, rather than `X`.
```
I don't think this functionality is compatible with the sequence regressor so it should just be removed. (If it is compatible, then it should have a test. But I feel like it's probably not compatible.)

MilesCranmer commented on pysr/sr.py:
```diff
@@ -2586,3 +2583,190 @@ def _mutate_parameter(param_name: str, param_value):
+        Parameters
+        ----------
+        X : ndarray | pandas.DataFrame
+            Training time series data of shape (n_samples, 1) or (1, n_samples).
```
```suggestion
            Training time series data of shape (n_times, 1).
```

MilesCranmer replied:
If this should have multiple features, I think it would be
```suggestion
            Training time series data of shape (n_times, n_features).
```
and the 1D scalar time series would just have `n_features=1`, but the user could pass multiple features if they wish.

MilesCranmer reviewed (commented):
General comment: should `PySRSequenceRegressor` allow for multiple features? Right now it is just a scalar feature over time. But what if you want to have two features over time, like `X_t` and `Z_t`, and get recursive information for both of them? 

(Just wondering why/why not)

MilesCranmer commented:
I think for `weights`, the time series approach should simply weight by time. So the weights should be shape `(n_times,)`. It should be one weight for every predicted value.

wenbang24 committed:
updated fit()

now uses super().fit()
also got rid of Xresampled

wenbang24 commented:
> I think for `weights`, the time series approach should simply weight by time. So the weights should be shape `(n_times,)`. It should be one weight for every predicted value.

yes I think that is how it currently works; I'll try and make the docstring a bit clearer

wenbang24 committed:
multidimensionality!!!

note: flattens all of the arrays

wenbang24 committed:
fixed variable names for multidimensionality

wenbang24 committed:
fixed variable names

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wenbang24 commented:
> General comment: should `PySRSequenceRegressor` allow for multiple features? Right now it is just a scalar feature over time. But what if you want to have two features over time, like `X_t` and `Z_t`, and get recursive information for both of them?
> 
> (Just wondering why/why not)

Yes, it does support multidimensional data, but since everything is flattened, the output looks unintuitive.

MilesCranmer commented on pysr/sr.py:
```diff
@@ -2586,3 +2583,122 @@ def _mutate_parameter(param_name: str, param_value):
+            Union[int, float, List[Union[int, float]]]
+        ] = None,
+        X_units: Optional[ArrayLike[str]] = None,
+    ) -> "PySRRegressor":
```
```suggestion
    ) -> "PySRSequenceRegressor":
```

MilesCranmer commented on pysr/sr.py:
```diff
@@ -2586,3 +2583,122 @@ def _mutate_parameter(param_name: str, param_value):
+        weights : ndarray | pandas.DataFrame
+            Weight array of the same shape as `X`, but not for the
+            first recurrence_history_length terms. Therefore, the shape is
+            (n_samples-recurrence_history_length, 1) or (1, n_samples-recurrence_history_length)
```
I don’t think the user should have to pass the shortened weights, the fit method should do this automatically from input `(n_samples, 1)`

MilesCranmer commented on pysr/sr.py:
```diff
@@ -2586,3 +2583,122 @@ def _mutate_parameter(param_name: str, param_value):
     return param_value
+
+
+class PySRSequenceRegressor(PySRRegressor):
```
Also, could this class be moved into a separate file? `sr.py` is too large as-is.

MilesCranmer commented on pysr/sr.py:
```diff
@@ -2586,3 +2583,122 @@ def _mutate_parameter(param_name: str, param_value):
+        **kwargs,
+    ):
+        self.recursive_history_length = recursive_history_length
+        super().__init__(**kwargs)
```
This line that calls the init of the parent class should go at the top of `__init__`, rather than the bottom.

MilesCranmer replied:
(Just a convention; doesn’t really matter here though)

MilesCranmer commented on pysr/sr.py:
```diff
@@ -2586,3 +2583,122 @@ def _mutate_parameter(param_name: str, param_value):
+        Parameters
+        ----------
+        X : ndarray | pandas.DataFrame
+            Training time series data of shape (n_times, ...).
```
Since PySRRegressor currently only handles 2D data of shape `(n_rows, n_features)`, this should be a fixed shape `(n_times, n_features)`

wenbang24 replied:
I added functionality to support any dimensional data :)

wenbang24 replied:
Currently the preprocessor is designed to give variables names like `x_20_16t_3`
This variable name means if you looked at the t-3 th data point, and looked at the index [20][16], you would find the variable
With 2d arrays we still need to concatenate and flatten

MilesCranmer replied:
Cool!

I guess I do think it should just force the user to flatten their data along the second axis though. Automatically flattening input data is a bit of a footgun in case the user provides images with different dimension than trained on. Better to force the user to be very explicit about the number of features they have. It's a slight inconvenience but it's much safer.

Basically, just have an assertion like

```python
assert len(X.shape) == 2
```

would be good enough. I think scikit-learn actually has some utility that does this automatically.

MilesCranmer commented on pysr/sr.py:
```diff
@@ -2586,3 +2583,122 @@ def _mutate_parameter(param_name: str, param_value):
+            if a custom `loss` was set, it will can be used
+            in arbitrary ways.
+        variable_names : list[str]
+            A list of names for the variables, rather than "xt_1", "xt_2", etc.
```
I think the variable names passed here should be for each feature. i.e., this should be equal to shape `(n_features,)`. Then the internal method should automatically turn that into `(n_features * recurrence_history_length,)`.

MilesCranmer commented on pysr/sr.py:
```diff
@@ -2586,3 +2583,122 @@ def _mutate_parameter(param_name: str, param_value):
+            a string representing a Julia expression. See DynamicQuantities.jl
+            https://symbolicml.org/DynamicQuantities.jl/dev/units/ for more
+            information.
+            Length should be equal to recurrence_history_length.
```
Length should be equal to `n_features`, right?

MilesCranmer commented on pysr/test/test.py:
```diff
@@ -513,6 +513,228 @@ def test_jl_function_error(self):
+            jl.seval("((::Val{x}) where x) -> x")(model.julia_options_.turbo), True
+        )
+
+    def test_sequence_multiline_seval(self):
```
I think this test was copied by mistake? It doesn’t seem relevant for sequence prediction

wenbang24 committed:
git is hard

wenbang24 committed:
added new preprocessing to predict

wenbang24 committed:
Merge branch 'recurrence' of github.com:wenbang24/PySR into recurrence

git is really hard

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wenbang24 committed:
removed unecessary test

wenbang24 committed:
small documentaion change

wenbang24 committed:
didn't the last commit work?

wenbang24 committed:
another small doc change

wenbang24 committed:
ok the preprocessing ACTUALLY works now

wenbang24 committed:
fixed custom variable names

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wenbang24 committed:
updated tests

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wenbang24 committed:
all tests passing!!!

but if test_sequence_weighted_bumper just run it again and it should be fine :)

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wenbang24 committed:
refactor: moved PySRSequenceRegressor to ssr.py

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wenbang24 commented:
oh is that bad?

wenbang24 commented:
I don't think the type checking is ever going to work :(
because PySRSequenceRegressor doesn't take y or y_units, but PySRRegressor does, so the signatures will always be different

wenbang24 committed:
slight change in multidimensional data error test

wenbang24 committed:
made the type checking work, added tests for unused variables

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wenbang24 committed:
yeah the override decorator wasn't needed

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wenbang24 commented:
nvm type checking works 🎉

wenbang24 committed:
this test is too harsh

MilesCranmer commented:
Cool! Yeah you can also generally use `cast` to tell the type checker what the type is supposed to be – it's good for getting around type issues that aren't actually real problems.

MilesCranmer commented on pysr/__init__.py:
```diff
@@ -8,6 +8,7 @@
 from .export_jax import sympy2jax
 from .export_torch import sympy2torch
 from .sr import PySRRegressor
+from .ssr import PySRSequenceRegressor
```
Can you give this a more descriptive filename? The `sr.py` is stuck with that name for historical reasons, but I would have preferred to call it `regressor.py`. So you could call the sequence one `regressor_sequence.py`?

MilesCranmer commented on pysr/ssr.py:
```diff
@@ -0,0 +1,171 @@
+        recursive_history_length: int = 0,
+        **kwargs,
+    ):
+        self.recursive_history_length = recursive_history_length
+        super().__init__(**kwargs)
```
```suggestion
        super().__init__(**kwargs)
        self.recursive_history_length = recursive_history_length
```

unless there's some reason to have it reversed? But it's standard practice to call the `super().__init__` at the top.

MilesCranmer commented on pysr/ssr.py:
```diff
@@ -0,0 +1,171 @@
+            Fitted estimator.
+        """
+
+        if y is not None:
+            warnings.warn(
+                "Recursive symbolic regression does not use `y` - this parameter is being ignored"
+            )
+        if y_units is not None:
+            warnings.warn(
+                "Recursive symbolic regression does not use `y_units` - this parameter is being ignored"
+            )
+        if Xresampled is not None:
+            warnings.warn(
+                "Recursive symbolic regression does not use `Xresampled` - this parameter is being ignored"
+            )
```
I think these should be errors.

MilesCranmer commented on pysr/ssr.py:
```diff
@@ -0,0 +1,171 @@
+                "Recursive symbolic regression does not use `Xresampled` - this parameter is being ignored"
+            )
+
+        if self.recursive_history_length <= 0:
+            raise ValueError(
+                "The `recursive_history_length` parameter must be greater than 0 (otherwise it's not recursion)."
+            )
+        if len(X.shape) > 2:
+            raise ValueError(
+                "Recursive symbolic regression only supports up to 2D data; please flatten your data first"
+            )
+        elif len(X) < 2:
+            raise ValueError(
+                "Recursive symbolic regression requires at least 2 datapoints; if you tried to pass a 1D array, use array.reshape(-1, 1)"
+            )
+        if len(X) <= self.recursive_history_length + 1:
+            raise ValueError(
+                f"Recursive symbolic regression with a history length of {self.recursive_history_length} requires at least {self.recursive_history_length + 2} datapoints."
+            )
+        if isinstance(weights, np.ndarray) and len(weights) != len(X):
+            raise ValueError("The length of `weights` must have shape (n_times,).")
```
Could all of these checks be moved to a separate function? You can see the `_check_assertions` function in `sr.py` as an example.

MilesCranmer commented on pysr/ssr.py:
```diff
@@ -0,0 +1,171 @@
+        if isinstance(weights, np.ndarray) and len(weights) != len(X):
+            raise ValueError("The length of `weights` must have shape (n_times,).")
+        y = X.copy()
+        temp = X.copy()[0]
```
More descriptive name of this? I'm not sure what this is otherwise. Also it would be preferable to copy only the 0th element, rather than taking the 0th element of a copy.

MilesCranmer commented on pysr/ssr.py:
```diff
@@ -0,0 +1,171 @@
+            raise ValueError("The length of `weights` must have shape (n_times,).")
+        y = X.copy()
+        temp = X.copy()[0]
+        X = np.lib.stride_tricks.sliding_window_view(
```
As a best practice, this `X` should have a different variable name. Generally you don't want to re-use variable names as it can introduce errors.

MilesCranmer commented on pysr/ssr.py:
```diff
@@ -0,0 +1,171 @@
+        X = np.lib.stride_tricks.sliding_window_view(
+            y[:-1].flatten(), self.recursive_history_length * temp.shape[0]
+        )[:: temp.shape[0], :]
+        y = np.array([i for i in y[self.recursive_history_length :]])
```
Same comment here. This `y` should have a different variable name.

MilesCranmer commented on pysr/ssr.py:
```diff
@@ -0,0 +1,171 @@
+        ValueError
+            Raises if the `best_equation` cannot be evaluated.
+        """
+        if len(X) < self.recursive_history_length:
+            raise ValueError(
+                f"Recursive symbolic regression with a history length of {self.recursive_history_length} requires at least {self.recursive_history_length} datapoints."
+            )
```
This could be handled by the same `_check_assertions` function requested above

MilesCranmer commented on pysr/ssr.py:
```diff
@@ -0,0 +1,171 @@
+                f"Recursive symbolic regression with a history length of {self.recursive_history_length} requires at least {self.recursive_history_length} datapoints."
+            )
+        temp = X.copy()
+        X = np.lib.stride_tricks.sliding_window_view(
```
Currently in `requirements.txt`, we have the constraint that

```
numpy>=1.13.0,<3.0.0
```

However, it looks like this function was only introduced in numpy 1.20.0: https://numpy.org/doc/stable/reference/generated/numpy.lib.stride_tricks.sliding_window_view.html. So could you also update the requirements to limit numpy to >=1.20.0?

MilesCranmer commented on pysr/ssr.py:
```diff
@@ -0,0 +1,171 @@
+        if isinstance(weights, np.ndarray):
+            weights = weights[self.recursive_history_length :]
+
+        if not variable_names:
+            if len(temp.shape) == 0:
+                variable_names = [
+                    f"xt_{i}" for i in range(self.recursive_history_length, 0, -1)
+                ]
+            elif len(temp.shape) == 1:
+                variable_names = [
+                    f"x{i}t_{j}"
+                    for j in range(temp.shape[0])
+                    for i in range(self.recursive_history_length, 0, -1)
+                ]
+        else:
+            variable_names = [
+                i + "t_" + str(j)
+                for i in variable_names
+                for j in range(self.recursive_history_length, 0, -1)
+            ]
```
Could this be split out to a separate function? Then it can be unittested without needing to call the entire `.fit` code.

MilesCranmer commented on pysr/ssr.py:
```diff
@@ -0,0 +1,171 @@
+        Parameters
+        ----------
+        X : ndarray | pandas.DataFrame
+            Training data of shape `(n_times, 1)`.
```
```suggestion
            Data of shape `(n_times, n_features)`.
```

MilesCranmer commented on pysr/ssr.py:
```diff
@@ -0,0 +1,171 @@
+
+        Returns
+        -------
+        x_predicted : ndarray of shape (n_samples, nout_)
```
```suggestion
        x_predicted : ndarray of shape (n_times, n_features)
```

MilesCranmer commented on pysr/ssr.py:
```diff
@@ -0,0 +1,171 @@
+        Raises
+        ------
+        ValueError
+            Raises if the `best_equation` cannot be evaluated.
```
Not sure what `best_equation` is here. Perhaps this was copied from a comment in `sr.py`? In any case I don't think `best_equation` exists now.

wenbang24 replied:
isn't it inherited from PySRRegressor?

MilesCranmer commented on pysr/test/test.py:
```diff
@@ -513,6 +513,282 @@ def test_jl_function_error(self):
         )
 
 
+class TestSequencePipeline(unittest.TestCase):
```
```suggestion
class TestSequenceRegressor(unittest.TestCase):
```

MilesCranmer commented on pysr/test/test.py:
```diff
@@ -513,6 +513,282 @@ def test_jl_function_error(self):
+        self.assertIn("exec", model.latex())
+
+    def test_sequence_multidimensional_data_error(self):
+        X = [
+            [
+                [1, 2],
+                [3, 4],
+                [5, 6],
+            ],
+            [
+                [7, 8],
+                [9, 10],
+                [11, 12],
+            ],
+            [
+                [13, 14],
+                [15, 16],
+                [17, 18],
+            ],
+        ]
+        for i in range(3, 5):
+            X.append(
+                [
+                    [X[i - 1][0][0] + X[i - 2][0][1], X[i - 1][0][1] + X[i - 2][1][0]],
+                    [X[i - 1][1][0] + X[i - 3][1][1], X[i - 1][1][1] + X[i - 3][1][0]],
+                    [X[i - 2][2][0] + X[i - 3][2][1], X[i - 2][2][1] + X[i - 3][2][0]],
+                ]
+            )
+        X = np.asarray(X)
```
```suggestion
        X = np.zeros((10, 1, 1))
```

Since it's just testing the presence of an error, can just do something like this

MilesCranmer commented on pysr/test/test.py:
```diff
@@ -513,6 +513,282 @@ def test_jl_function_error(self):
+            jl.seval("((::Val{x}) where x) -> x")(model.julia_options_.bumper), True
+        )
+
+    def test_sequence_multiprocessing_turbo_custom_objective(self):
+        X = [1]
+        for i in range(1, 20):
+            X.append(np.sqrt(X[i - 1]) + 1)
+        X = np.asarray(X).reshape(-1, 1)
+        model = PySRSequenceRegressor(
+            **self.default_test_kwargs,
+            # Turbo needs to work with unsafe operators:
+            unary_operators=["sqrt"],
+            procs=2,
+            multithreading=False,
+            turbo=True,
+            early_stop_condition="stop_if(loss, complexity) = loss < 1e-10 && complexity == 1",
+            loss_function="""
+            function my_objective(tree::Node{T}, dataset::Dataset{T}, options::Options) where T
+                prediction, flag = eval_tree_array(tree, dataset.X, options)
+                !flag && return T(Inf)
+                abs3(x) = abs(x) ^ 3
+                return sum(abs3, prediction .- dataset.y) / length(prediction)
+            end
+            """,
+        )
+        model.fit(X)
+        print(model.equations_)
+        best_loss = model.equations_.iloc[-1]["loss"]
+        self.assertLessEqual(best_loss, 1e-10)
+        self.assertGreaterEqual(best_loss, 0.0)
+
+        # Test options stored:
+        self.assertEqual(
+            jl.seval("((::Val{x}) where x) -> x")(model.julia_options_.turbo), True
+        )
```
Not sure what this test is probing differently from the one in `TestPipeline`? The sequence regressor tests should only test parts unique to the sequence regressor, rather than testing all of the options again.

MilesCranmer commented on pysr/ssr.py:
```diff
@@ -0,0 +1,171 @@
+        super().__init__(**kwargs)
+
+    def fit(
+        self,
+        X,
+        y=None,
+        Xresampled=None,
+        weights=None,
+        variable_names: Optional[ArrayLike[str]] = None,
+        complexity_of_variables: Optional[
+            Union[int, float, List[Union[int, float]]]
+        ] = None,
+        X_units: Optional[ArrayLike[str]] = None,
+        y_units=None,
```
```suggestion
        self,
        X,
        *,
        weights=None,
        variable_names: Optional[ArrayLike[str]] = None,
        complexity_of_variables: Optional[
            Union[int, float, List[Union[int, float]]]
        ] = None,
        X_units: Optional[ArrayLike[str]] = None,
```

(This `*` will be nice here for preventing potential errors.)

wenbang24 replied:
mypy doesn't like it :(
but I will use that in future projects 👍

wenbang24 committed:
swap super.__init__ and other thing in __init__

Co-authored-by: Miles Cranmer <miles.cranmer@gmail.com>

wenbang24 committed:
now needs numpy 1.20

wenbang24 committed:
changed predict docstring

wenbang24 committed:
change test sequence name

wenbang24 committed:
change multidimensional data error test

wenbang24 committed:
renamed ssr.py to regressor_sequence.py

wenbang24 committed:
moved assertions to new function, fixed error in variable name generation

wenbang24 committed:
removed need for temp

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wenbang24 committed:
changed variable names and made target generation a bit more efficient

wenbang24 committed:
made predict use _check_assertions as well

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wenbang24 committed:
moved variable name generation to new function and also added variable name tests

wenbang24 committed:
remove unnecessary test

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

MilesCranmer commented on pysr/test/test.py:
```diff
@@ -513,6 +513,257 @@ def test_jl_function_error(self):
+        model.fit(X, variable_names=["c1"])
+        self.assertIn("c1t_1", model.equations_.iloc[-1]["equation"])
+
+    def test_sequence_weighted_bumper(self):
+        X = [1, 1, 1]
+        for i in range(3, 30):
+            X.append(X[i - 1] + X[i - 2] + X[i - 3])
+        X = np.asarray(X).reshape(-1, 1)
+        weights = np.ones_like(X).reshape(-1)
+        model = PySRSequenceRegressor(
+            **self.default_test_kwargs,
+            early_stop_condition="stop_if(loss, complexity) = loss < 1e-4 && complexity <= 5",
+            bumper=True,
+        )
+        model.fit(X, weights=weights)
+        print(model.equations_)
+        self.assertLessEqual(model.get_best()["loss"], 1e-1)
+        self.assertEqual(
+            jl.seval("((::Val{x}) where x) -> x")(model.julia_options_.bumper), True
+        )
+
```
```suggestion
```

MilesCranmer commented on pysr/test/test.py:
```diff
@@ -513,6 +513,257 @@ def test_jl_function_error(self):
+            jl.seval("((::Val{x}) where x) -> x")(model.julia_options_.bumper), True
+        )
+
+    def test_sequence_high_precision_search_custom_loss(self):
+        X = [1, 1, 1]
+        for i in range(3, 30):
+            X.append(X[i - 1] + X[i - 2] + X[i - 3])
+        X = np.asarray(X).reshape(-1, 1)
+        model = PySRSequenceRegressor(
+            **self.default_test_kwargs,
+            early_stop_condition="stop_if(loss, complexity) = loss < 1e-4 && complexity == 3",
+            elementwise_loss="my_loss(prediction, target) = (prediction - target)^2",
+            precision=64,
+            parsimony=0.01,
+            warm_start=True,
+        )
+        model.fit(X)
+
+        # We should have that the model state is now a Float64 hof:
+        test_state = model.raw_julia_state_
+        self.assertTrue(jl.typeof(test_state[1]).parameters[1] == jl.Float64)
+
+        # Test options stored:
+        self.assertEqual(
+            jl.seval("((::Val{x}) where x) -> x")(model.julia_options_.turbo), False
+        )
+
```
```suggestion
```

MilesCranmer commented on pysr/test/test.py:
```diff
@@ -513,6 +513,257 @@ def test_jl_function_error(self):
+            "number of elements in `complexity_of_variables`", str(cm.exception)
+        )
+
+    def test_sequence_error_message_both_variable_complexity(self):
+        X = [1, 1]
+        for i in range(2, 100):
+            X.append(X[i - 1] + X[i - 2])
+        X = np.asarray(X).reshape(-1, 1)
+        model = PySRSequenceRegressor(
+            **self.default_test_kwargs, complexity_of_variables=[1, 2]
+        )
+        with self.assertRaises(ValueError) as cm:
+            model.fit(X, complexity_of_variables=[1, 2, 3])
+
+        self.assertIn(
+            "You cannot set `complexity_of_variables` at both `fit` and `__init__`.",
+            str(cm.exception),
+        )
+
```
```suggestion
```

MilesCranmer commented on pysr/test/test.py:
```diff
@@ -513,6 +513,257 @@ def test_jl_function_error(self):
+            str(cm.exception),
+        )
+
+    def test_sequence_warm_start_set_at_init(self):
+        # Smoke test for bug where warm_start=True is set at init
+        X = [1, 1, 1]
+        for i in range(3, 30):
+            X.append(X[i - 1] + X[i - 2] + X[i - 3])
+        X = np.asarray(X).reshape(-1, 1)
+        regressor = PySRSequenceRegressor(
+            **self.default_test_kwargs, warm_start=True, max_evals=10
+        )
+        regressor.fit(X)
+
```
```suggestion
```

MilesCranmer commented on pysr/test/test.py:
```diff
@@ -513,6 +513,257 @@ def test_jl_function_error(self):
+        )
+        regressor.fit(X)
+
+    def test_sequence_noisy_builtin_variable_names(self):
+        X = [1, 1]
+        for i in range(2, 30):
+            X.append(X[i - 1] + X[i - 2])
+        X = np.asarray(X).reshape(-1, 1)
+        model = PySRSequenceRegressor(
+            binary_operators=["+"],
+            **self.default_test_kwargs,
+            early_stop_condition="stop_if(loss, complexity) = loss < 0.05 && complexity == 2",
+        )
+        # We test builtin variable names
+        model.fit(X, variable_names=["exec"])
+        self.assertLessEqual(model.get_best()["loss"], 1e-2)
+        self.assertLessEqual(model.get_best()["loss"], 1e-2)
+        self.assertIn("exec", model.latex())
+
```
```suggestion
```

MilesCranmer reviewed (commented):
Some of these tests aren't necessary; they don't test unique aspects of the sequence regressor. The sequence regressor tests should probe the sequence regressor code itself

MilesCranmer commented:
Btw you can do `pre-commit install` to have the precommit checks run locally rather than as additional git commits.

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,186 @@
+        raise ValueError(
+            "The length of `X_units` must be equal to the number of features in `X`."
+        )
+    return (X, recursive_history_length, weights, variable_names, X_units)
```
```suggestion
```

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,186 @@
+            Fitted estimator.
+        """
+
+        if y is not None:
+            warnings.warn(
+                "Recursive symbolic regression does not use `y` - this parameter is being ignored"
+            )
+        if y_units is not None:
+            warnings.warn(
+                "Recursive symbolic regression does not use `y_units` - this parameter is being ignored"
+            )
+        if Xresampled is not None:
+            warnings.warn(
+                "Recursive symbolic regression does not use `Xresampled` - this parameter is being ignored"
+            )
+
```
Move these to `_check_assertions` (and convert them to errors rather than warnings)

wenbang24 committed:
made unused variables throw errors

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wenbang24 committed:
changed check_assertions to have no return value

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,189 @@
+                for i in variable_names
+            ]
+
+    def fit(
+        self,
+        X,
+        y=None,
+        Xresampled=None,
+        weights=None,
+        variable_names: Optional[ArrayLike[str]] = None,
+        complexity_of_variables: Optional[
+            Union[int, float, List[Union[int, float]]]
+        ] = None,
+        X_units: Optional[ArrayLike[str]] = None,
+        y_units=None,
+    ) -> "PySRSequenceRegressor":
```
```suggestion
    def fit(
        self,
        X,
        *,
        weights=None,
        variable_names: Optional[ArrayLike[str]] = None,
        complexity_of_variables: Optional[
            Union[int, float, List[Union[int, float]]]
        ] = None,
        X_units: Optional[ArrayLike[str]] = None,
    ) -> "PySRSequenceRegressor":
```

(Will also require updating the other code)

wenbang24 replied:
mypy still says `Signature of "fit" incompatible with supertype "PySRRegressor"`

MilesCranmer replied:
Ah, I see... Hm. Maybe mypy is pointing to a design issue since the classes are fundamentally different. In other words, maybe the PySRRegressor shouldn't actually be a supertype, but rather should be an **attribute** of PySRSequenceRegressor?

Something like

```julia
class PySRSequenceRegressor(BaseEstimator):
    def __init__(recursive_history_length, **kwargs):
        self._regressor = PySRRegressor(**kwargs)
```


and then call `self._regressor...` instead of `super()...`?

wenbang24 committed:
removed unecessary tests

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,189 @@
+            y_units,
+        )
+
+        y = X[self.recursive_history_length :]
+        newX = np.lib.stride_tricks.sliding_window_view(
+            X[:-1].flatten(), self.recursive_history_length * X.shape[1]
+        )[:: y.shape[1], :]
```
```suggestion
        current_X = X[self.recursive_history_length :]
        historical_X = np.lib.stride_tricks.sliding_window_view(
            X[:-1].flatten(), self.recursive_history_length * X.shape[1]
        )[:: current_X.shape[1], :]
```

I think these are more suitable variable names. Wdyt?

wenbang24 replied:
yeah I think its good (I need to get better at naming variables) but aren't they supposed to be the other way around?
bcuz the stride tricks function is significantly changing the original X, so the output should be current_X

MilesCranmer replied:
> so the output should be current_X

I think that's what's suggested here, no?

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,189 @@
+        variable_names = self._variable_names(y, variable_names)
+
+        super().fit(
+            newX,
+            y,
+            weights=weights,
```
```suggestion
            X=historical_X,
            y=current_X,
            weights=weights,
```

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,189 @@
+            Raises if the `best_equation` cannot be evaluated.
+        """
+        X = _check_assertions(X, self.recursive_history_length)[0]
+        X = np.lib.stride_tricks.sliding_window_view(
+            X.flatten(), self.recursive_history_length * np.prod(X.shape[1])
+        )[:: X.shape[1], :]
+        return super().predict(X, index=index)
```
I think `predict` should output something that is the same shape as `X`. To do this, I think the predictions should be stacked with a matrix of `NaN`s for rows which do not have enough history to predict them.

wenbang24 replied:
currently using np.nan and it throws `TypeError: only integer scalar arrays can be converted to a scalar index`
also what would the model output for the rows with NaN's?

MilesCranmer replied:
Not sure what that error is. Maybe concatenating incorrectly or something?

MilesCranmer replied:
The model should automatically output NaNs for those I think.

wenbang24 replied:
```py
File "/opt/homebrew/lib/python3.12/site-packages/sklearn/utils/validation.py", line 172, in _assert_all_finite_element_wise
    raise ValueError(msg_err)
ValueError: Input contains NaN.
```
maybe not?
not much we can do about that

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,189 @@
+        ValueError
+            Raises if the `best_equation` cannot be evaluated.
+        """
+        X = _check_assertions(X, self.recursive_history_length)[0]
```
```suggestion
        _check_assertions(X, self.recursive_history_length)
```

wenbang24 committed:
changed up variable names

wenbang24 committed:
fixed up check_assertions in predict

wenbang24 committed:
fixed bug in variable name generation

wenbang24 committed:
fixed bug in assertion checking in predict

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,189 @@
+            y_units,
+        )
+
+        historical_X = X[self.recursive_history_length :]
+        current_X = np.lib.stride_tricks.sliding_window_view(
+            X[:-1].flatten(), self.recursive_history_length * X.shape[1]
+        )[:: historical_X.shape[1], :]
```
Pretty sure these names are reversed. `historical_X` should be switched with `current_X`

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,189 @@
+            Raises if the `best_equation` cannot be evaluated.
+        """
+        _check_assertions(X, self.recursive_history_length)
+        current_X = np.lib.stride_tricks.sliding_window_view(
```
We are _predicting_ the `current_X` based on historical data so this should be `historical_X`

wenbang24 committed:
changed up variable names

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,192 @@
+        super().__init__(**kwargs)
+        self.recursive_history_length = recursive_history_length
+
+    def _variable_names(self, y, variable_names=None):
```
```suggestion
    def _construct_variable_names(self, y, variable_names=None):
```

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,192 @@
+
+    def _variable_names(self, y, variable_names=None):
+        if not variable_names:
+            if y.shape[1] == 1:
```
Since only `y.shape[1]` is used, it would be easier and more readable to pass `n_features: int` here rather than all of `y` as an argument.

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,192 @@
+        raise ValueError("Recursive symbolic regression does not use `Xresampled`")
+
+
+class PySRSequenceRegressor(PySRRegressor):
```
Needs class docstring

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,192 @@
+        Parameters
+        ----------
+        X : ndarray | pandas.DataFrame
+            Training time series data of shape (n_times, n_features).
```
```suggestion
            Time series training data of shape (n_times, n_features).
```

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,192 @@
+        Parameters
+        ----------
+        X : ndarray | pandas.DataFrame
+            Training time series data of shape (n_times, n_features).
```
Might also be nice to allow input of shape `(n_times,)`. This isn't allowed in the regular `PySRRegressor`, but only because a user might pass a single row of data with multiple features. But here, there is no ambiguity: you would always want to pass multiple times. So we can allow 1D input.

MilesCranmer replied:
Although I guess `self._validate_data` might require 2D... Hm...

wenbang24 replied:
if it's 1D, can we just make it 2D with reshape(-1, 1) before calling _validate_data?

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,192 @@
+        self : object
+            Fitted estimator.
+        """
+
```
```suggestion
        X = self._validate_data(X)
```

This is a method from the base class `sklearn.BaseEstimator` which will automatically convert `X` to a numpy ndarray if it happens to be a pandas dataframe or pure python list.

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,192 @@
+        ValueError
+            Raises if the `best_equation` cannot be evaluated.
+        """
+        _check_assertions(X, recursive_history_length=self.recursive_history_length)
```
```suggestion
        X = self._validate_data(X)
        _check_assertions(X, recursive_history_length=self.recursive_history_length)
```

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,192 @@
+                for i in variable_names
+            ]
+
+    def fit(
```
Right now, if you pass `X` with multiple features, it will try to fit a recurrence relation for every one. But, I could see it being useful to pass multiple features, and only predict a subset of them. What do you think? Should that be implemented?

MilesCranmer replied:
Actually I think this would add complication and not really be worth it. So I take back this suggestion.

MilesCranmer commented:
For the `predict` step I am also wondering if it should recursively roll out the prediction up to some user-specified time in the future? Basically, if the user provides the sequence from `1` to `n`, for a history length `h`, and asks for an extra `m` steps, the result would be times `h` through `n+m`? Right now we only predict up to `n` which doesn't really make sense because the user has already provided that time point.

wenbang24 committed:
changed name of variable name generator

wenbang24 committed:
changed variable name generator to take n_features

wenbang24 committed:
Updated docstring

wenbang24 committed:
added validation of X

wenbang24 committed:
added another validation

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wenbang24 commented:
oh my bad I was not supposed to overwrite example.py

wenbang24 committed:
added doc string for PySRSymbolicRegressor

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wenbang24 committed:
changed variable name generation function to isinstance rather than truthy bcuz python gets mad :(

wenbang24 committed:
think i fixed predicting shape

wenbang24 committed:
changed PySRSequenceRegressor to inherit from BaseEstimator and have PySRRegressor as an attribute

wenbang24 commented:
> For the `predict` step I am also wondering if it should recursively roll out the prediction up to some user-specified time in the future? Basically, if the user provides the sequence from `1` to `n`, for a history length `h`, and asks for an extra `m` steps, the result would be times `h` through `n+m`? Right now we only predict up to `n` which doesn't really make sense because the user has already provided that time point.

maybe instead of doing the padding with NaNs thing, we can do this?
also we should probably let user pick how many times they want to extrapolate

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wenbang24 committed:
padding with NaNs does not work

MilesCranmer commented:
Yep sounds good

wenbang24 committed:
made extrapolating when prredicting work :)

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wenbang24 commented:
bad news
since PySRRegressor is now an attribute of PySRSequenceRegressor all of the attributes like `model.equations_` and functions like `model.get_best()` don't work anymore :(
do we revert back to inheriting from PySRRegressor or is there some workaround?

wenbang24 committed:
updated dosctring to have extra_preditciotns

wenbang24 commented:
woah delegation is pretty cool
but I keep getting `sklearn.exceptions.NotFittedError: This PySRRegressor instance is not fitted yet. Call 'fit' with appropriate arguments before using this estimator.`

wenbang24 commented:
for some reason 
```
self._regressor.__repr__ = <bound method PySRRegressor.__repr__ of PySRRegressor.equations_ = [
[
           pick  score       equation       loss  complexity
        0          0.0          x1t_3  72.666664           1
        1  >>>>    inf  x0t_2 + x0t_1   0.000000           3
], [
           pick  score       equation       loss  complexity
        0          0.0          x1t_2  130.66667           1
        1  >>>>    inf  x1t_1 + x1t_2    0.00000           3
]]>
```

wenbang24 commented:
btw this is the code
```py
def __getattr__(self, name):
    #return self._regressor.__getattr__(name)
    try:
        return getattr(self._regressor, name)
    except AttributeError:
        raise AttributeError(f"'{self.__class__.__name__}' object has no attribute '{name}'")

def __delattr__(self, name):
    if name == "_regressor":
        print("no delete regressor")
    else:
        delattr(self._regressor, name)

def __setattr__(self, name, value) -> None:
    if name == "_regressor":
        object.__setattr__(self, name, value)
    else:
        setattr(self._regressor, name, value)
        print(self._regressor.__dict__[name])
```

wenbang24 committed:
tried delegation but this doesn't work

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,257 @@
+            return output.reshape(-1, X.shape[1])
+        return pred
+
+    def __getattr__(self, name):
+        # return self._regressor.__getattr__(name)
+        try:
+            return getattr(self._regressor, name)
+        except AttributeError:
+            raise AttributeError(
+                f"'{self.__class__.__name__}' object has no attribute '{name}'"
+            )
+
+    def __delattr__(self, name):
+        if name == "_regressor":
+            print("no delete regressor")
+        else:
+            delattr(self._regressor, name)
+
+    def __setattr__(self, name, value) -> None:
+        if name == "_regressor":
+            object.__setattr__(self, name, value)
+        else:
+            setattr(self._regressor, name, value)
+            print(self._regressor.__dict__[name])
```
I think we should just define each method individually and forward arguments, like

```python
    def sympy(self, index=None):
        return self._regressor.sympy(index=index)
```

This is much safer, as it forces us to think about which functions are compatible and which are not. For example I think `.pytorch` and `.jax` are not compatible (and it's not worth it to get them working), so we should just leave those unimplemented.

wenbang24 replied:
how do I forward properties like `equations_`?

MilesCranmer replied:
```
@property
def equations_
    return self._regressor.equations_
```

wenbang24 committed:
forwarded all methods and properties from PySRRegressor to PySRSequecneRegressor

wenbang24 commented:
all tests passing except for `test_sequence_variable_names`
it says truth value of an array with more than one element is ambiguous for `n_features == 1` which is weird because it should be an integer because of the type hinting: `def _construct_variable_names(self, n_features: int, variable_names=None)`

wenbang24 commented:
are there any other changes that need to be made?

wenbang24 committed:
finally fixed the bug in variable name test

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,381 @@
+    def __repr__(self):
+        return self._regressor.__repr__()
+
+    def __getstate__(self):
+        return self._regressor.__getstate__()
```
This is incorrect as it doesn't include the top-level model's state

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,381 @@
+            **pysr_kwargs,
+        )
+    
+    def __repr__(self):
+        return self._regressor.__repr__()
```
This prints `PySRRegressor(...)` which is incorrect

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,381 @@
+    def get_hof(self):
+        return self._regressor.get_hof()
+    
+    def latex_table(
+        self,
+        indices=None,
+        precision=3,
+        columns=["equation", "complexity", "loss", "score"],
+    ):
+        return self._regressor.latex_table(indices=indices, precision=precision, columns=columns)
+    
```
These are incorrect since the latex has `y=...`. But `y` is not defined in the sequence regressor

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,381 @@
+    def X_units_(self):
+        return self._regressor.X_units_
+
+    @property
+    def y_units_(self):
+        return self._regressor.y_units_
+
```
Does not exist in the sequence regressor

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,381 @@
+            instead of `variable_names`. Cannot contain spaces or special
+            characters. Avoid variable names which are also
+            function names in `sympy`, such as "N".
+            The number of variable names must be equal to (n_features,).
```
Missing `complexity_of_variables` in the docstring

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,381 @@
+            y_units=y_units,
+            complexity_of_variables=complexity_of_variables,
+        )
+        self._regressor.__dict__["__sklearn_is_fitted__"] = True
+        self._regressor.__dict__["selection_mask_"] = self._regressor.selection_mask_
+        self._regressor.__dict__["feature_names_in_"] = (
+            self._regressor.feature_names_in_
+        )
+        self._regressor.__dict__["nout_"] = self._regressor.nout_
```
Why do these use `__dict__` here instead of other ways of setting these?

MilesCranmer replied:
And also why are each of these needed? I think some of them are internal properties of `PySRRegressor`. But if they're not needed within the sequence regressor there's likely no reason to include them.

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,381 @@
+            y_units=y_units,
+            complexity_of_variables=complexity_of_variables,
+        )
+        self._regressor.__dict__["__sklearn_is_fitted__"] = True
```
Why is this needed?

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,381 @@
+            Union[int, float, List[Union[int, float]]]
+        ] = None,
+        X_units: Optional[ArrayLike[str]] = None,
+        y_units=None,
```
```suggestion
```

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,381 @@
+    def fit(
+        self,
+        X,
+        y=None,
```
```suggestion
```

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,381 @@
+    def fit(
+        self,
+        X,
+        y=None,
+        Xresampled=None,
```
```suggestion
```

MilesCranmer replied:
```suggestion
```

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,381 @@
+        y_units=None,
+    ) -> "PySRSequenceRegressor":
+        """
+        Search for equations to fit the time series dataset and store them in `self.equations_`.
```
Instead of "time series" could it be called a "sequence" or something like that? I think it is more general than just time series datasets. You could have mathematical recurrence relationships for example – there's no sense of "time" there

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,381 @@
+        self,
+        recursive_history_length: int = 0,
+        **kwargs,
+    ):
+        self._regressor = PySRRegressor(**kwargs)
```
Still needs the `super().__init__()`

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,381 @@
+        raise ValueError("Recursive symbolic regression does not use `Xresampled`")
+
+
+class PySRSequenceRegressor(BaseEstimator):
```
Are there other sklearn base classes to use here? Or do we only want `BaseEstimator`?

wenbang24 replied:
well PySRRegressor inherits from MultiOutputMixin and RegressorMixin so we could use those

MilesCranmer replied:
I'm not sure it fits those though, maybe read about the requirements for those abstract classes? i.e., the scikit-learn docs.

wenbang24 replied:
I think it does fit RegressorMixin

> Mixin class for all regression estimators in scikit-learn.

wenbang24 replied:
I found one line of documentation
PySRSequenceRegressor does require multi output for multidimensional recurrent sequences, right?
> Mixin to mark estimators that support multioutput.

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,381 @@
+        )
+    if y is not None:
+        raise ValueError("Recursive symbolic regression does not use `y`")
+    if y_units is not None:
+        raise ValueError("Recursive symbolic regression does not use `y_units`")
+    if Xresampled is not None:
+        raise ValueError("Recursive symbolic regression does not use `Xresampled`")
+
```
These aren't needed; can just remove the arguments instead.

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,381 @@
+    def equations_(self):
+        return self._regressor.equations_
+
+    # this causes errors
+    """ @property
+    def n_features_in_(self):
+        return self._regressor.n_features_in_ """
+
```
I think sklearn needs to set this itself, rather than the class, no?

wenbang24 committed:
removed unecessary variables

wenbang24 committed:
updated docstring

wenbang24 committed:
added __getstate__ to PySRSequenceRegressor

wenbang24 committed:
update docstring

wenbang24 committed:
added super().__init()

wenbang24 committed:
removed n_features_in

wenbang24 commented on example.py:
```diff
@@ -1,27 +1,20 @@
+print(model._regressor.__dict__)
+print(model._regressor.__repr__())
+print(hasattr(model._regressor, "feature_names_in_"))
+print(hasattr(model._regressor, "selection_mask_"))
+print(hasattr(model._regressor, "nout_"))
 
-model = PySRRegressor(
-    model_selection="best",  # Result is mix of simplicity+accuracy
-    niterations=40,
```
yeah I was just gonna test out PySRSequenceRegressor to pinpoint an error in variable names
but I accidentally committed it 🤦

wenbang24 committed:
example.py is back

wenbang24 committed:
variable names for unnamed 1D sequences work now

wenbang24 committed:
latex table no longer says y, but xt_0 (or whatever the variable name is)

wenbang24 committed:
added docstring for complexity of variables

wenbang24 committed:
grammar is hard

wenbang24 committed:
remove unused imports

wenbang24 committed:
uncomment tests

wenbang24 committed:
newlines??

wenbang24 committed:
put a comment back

wenbang24 committed:
new inherits

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,461 @@
+    def __repr__(self):
+        return self._regressor.__repr__().replace("PySRRegressor", "PySRSequenceRegressor")
+
+    def __getstate__(self):
```
I don't think this function is needed? The built-in `__getstate__` should already work. The only reason PySRRegressor implements a custom one is because some of its attributes are not pickleable. But `__getstate__` works recursively so should just work fine.

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,461 @@
+    @property
+    def X_units_(self):
+        return self._regressor.X_units_
+        return self._regressor.y_units_
```
```suggestion
```

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,461 @@
+    @property
+    def complexity_of_variables_(self):
+        return self._regressor.complexity_of_variables_
+
+    @property
+    def X_units_(self):
+        return self._regressor.X_units_
+        return self._regressor.y_units_
+
+    @property
+    def nout_(self):
+        return self._regressor.nout_
+
+    @property
+    def selection_mask_(self):
+        return self._regressor.selection_mask_
+
+    @property
+    def tempdir_(self):
+        return self._regressor.tempdir_
+
+    @property
+    def equation_file_(self):
+        return self._regressor.equation_file_
+
+    @property
+    def julia_state_stream_(self):
+        return self._regressor.julia_state_stream_
+
+    @property
+    def julia_options_stream_(self):
+        return self._regressor.julia_options_stream_
+
+    @property
+    def equation_file_contents_(self):
+        return self._regressor.equation_file_contents_
+
+    @property
+    def show_pickle_warnings_(self):
+        return self._regressor.show_pickle_warnings_
```
```suggestion
```
Not needed; it's an internal property

MilesCranmer replied:
Apart from `equations_`, what are the other ones needed for?

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,461 @@
+            return output.reshape(-1, X.shape[1])
+        return pred
+
+    def from_file(
```
I don't think this method will work the way it is written. The method you are overloading is a `@classmethod` which works a bit differently from a normal Python method.

MilesCranmer commented:
I think as a general guiding principle for this PR, since it implements a lot of new methods, we should try to have the testing get to 100% code coverage on the git diff. If we are at less than 100% diff code coverage then it points to things that we can take out (we should prefer less code rather than more... makes my job easier). For example, most of the `@property` methods are not needed – with code coverage we would be able to see that automatically (which can be done without needing my review so often).

wenbang24 committed:
remove y units

wenbang24 committed:
remove show pickel warnings

wenbang24 committed:
removed a lot of unneccesary properties

wenbang24 committed:
removed __getstate__

wenbang24 committed:
removed some commmented out code

wenbang24 committed:
since recursive history length >= 1, we don't need to test for len(X) < 2 bcuz recursive_history_length covers that

wenbang24 committed:
more tests

wenbang24 committed:
update dosctring

wenbang24 committed:
more tests

wenbang24 committed:
more tests

wenbang24 commented:
so I rewrote (copied) from_file(), but it doesn't work
`ValueError: Length mismatch: Expected axis has 1 elements, new values have 2 elements`
ill try and resolve this in the afternoon

wenbang24 commented:
I think PySRSequenceRegressor might have to pickle recursive history length to resolve that value error - pandas thinks there should only be 1 input feature but because recursive history length is 2, there are 2x1 = 2 input features which messes stuff up
alternatively we could change n_features_in and nout manually?

wenbang24 committed:
Merge branch 'master' into recurrence

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wenbang24 committed:
ok i think from_file works now

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wenbang24 committed:
so i think we need to pickle recursive history length :(

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wenbang24 committed:
remove debug print

wenbang24 committed:
ok fromfile actually works now

wenbang24 committed:
added weight test

wenbang24 committed:
added feature name and selection mask tests

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wenbang24 commented:
ok I don't think its really worth pickling recurrence history length

wenbang24 committed:
add recursive history length parameter

wenbang24 committed:
fixed typing for variable names in fit

wenbang24 commented:
I don't think my coverage thing is working; how do I check coverage with coveralls?

MilesCranmer commented:
For coverage I usually use:

```bash
coverage run -m pysr test main,cli,startup
```

You need to install the `coverage` python package

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,425 @@
+        )
+
+
+class PySRSequenceRegressor(MultiOutputMixin, RegressorMixin, BaseEstimator):
```
What do `MultiOutputMixin` and `RegressorMixin` add to the methods? I'm not sure it's compatible with them since they have different signatures.

wenbang24 replied:
oh true this is wrapper of PySRRegressor

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,425 @@
+        return model
+
+    def __repr__(self):
+        return self._regressor.__repr__().replace(
+            "PySRRegressor", "PySRSequenceRegressor"
+        )
+
```
Is there a cleaner way to do this? `replace` replaces **all** occurrences so this might cause some issues if PySRRegressor appears as a variable or shows up somewhere else in the string for whatever reason

wenbang24 replied:
since __repr__ in PySRRegressor starts with "PySRRegressor.equations_ = [" I changed it to only replace the first occurrence :)

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,425 @@
+        return pred
+
+    @classmethod
+    def from_file(
```
I think it would be much simpler to just call `PySRRegressor.from_file`, and use that to instantiate `_regressor` in a new `PySRSequenceRegressor`

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,425 @@
+        equation_file: PathLike,
+        *pysr_args,
+        recursive_history_length: int,
+        binary_operators: Optional[List[str]] = None,
+        unary_operators: Optional[List[str]] = None,
+        n_features_in: Optional[int] = None,
+        feature_names_in: Optional[ArrayLike[str]] = None,
+        selection_mask: Optional[NDArray[np.bool_]] = None,
+        nout: int = 1,
+        **pysr_kwargs,
```
Should just have a `**kwargs` and forward them `PySRRegressor.from_file`

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,425 @@
+        """
+        self._regressor.refresh()
+
+        if isinstance(self._regressor.equations_, list):
+            if indices is not None:
+                assert isinstance(indices, list)
+                assert isinstance(indices[0], list)
+                assert len(indices) == self._regressor.nout_
+            variable_names = self.variable_names
+            if variable_names is not None:
+                variable_names = [
+                    variable_name + "t_0" for variable_name in variable_names
+                ]
+            else:
+                variable_names = [f"x{i}t_0" for i in range(self.n_features)]
+            table_string = sympy2multilatextable(
+                self._regressor.equations_,
+                indices=indices,
+                precision=precision,
+                columns=columns,
+                output_variable_names=variable_names,
+            )
+        elif isinstance(self.equations_, pd.DataFrame):
+            if indices is not None:
+                assert isinstance(indices, list)
+                assert isinstance(indices[0], int)
+            if self.variable_names is not None:
+                assert len(self.variable_names) == 1
+                variable_name = self.variable_names[0] + "t_0"
+            else:
+                variable_name = "xt_0"
+            table_string = sympy2latextable(
+                self._regressor.equations_,
+                indices=indices,
+                precision=precision,
+                columns=columns,
+                output_variable_name=variable_name,
+            )
+        else:
```
Is all of this code needed? Perhaps there's a way to simply rely on `_regressor.latex_table`? 

Basically, you should try to write as little code as possible where possible. Makes maintenance (for me) much easier.

wenbang24 replied:
unfortunately not because of the variable names
but should I change up latex_table in PySRRegressor to accept a variable names parameter or just keep it like it is

MilesCranmer replied:
Oh why doesn’t passing the normal variable names work (i.e., having PySRRegressir store them automatically after running .fit?)

wenbang24 replied:
actually im gonna do that

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,425 @@
+            "PySRRegressor", "PySRSequenceRegressor"
+        )
+
+    @property
+    def julia_options_(self):
+        return self._regressor.julia_options_
+
+    @property
+    def julia_state_(self):
+        return self._regressor.julia_state_
+
```
```suggestion
```
These aren't used anywhere as far as I can tell so can be deleted

MilesCranmer commented on pysr/test/test.py:
```diff
@@ -1271,13 +1573,14 @@ def test_unit_propagation(self):
-        TestHelpMessages,
-        TestLaTeXTable,
-        TestDimensionalConstraints,
+        # TestPipeline,
+        TestSequenceRegressor,
+        # TestBest,
+        # TestFeatureSelection,
+        # TestMiscellaneous,
+        # TestHelpMessages,
+        # TestLaTeXTable,
+        # TestDimensionalConstraints,
```
I think this was committed on accident

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,425 @@
+        **kwargs,
+    ):
+        self._regressor = PySRRegressor(**kwargs)
+        super().__init__()
```
Why is `super().__init__()` in the middle of these two lines?

MilesCranmer commented:
> I think PySRSequenceRegressor might have to pickle recursive history length

This should be automatic though, right? `__getstate__` normally just saves all parameters from `init`, recursively calling `__getstate__` on all linked objects. No need for a custom method.

MilesCranmer reviewed (changes requested):
Left some comments throughout. I think the implementation complexity is still a bit high so I would try to simplify things and rely on existing code in PySRRegressor where possible. Let me know when the PR is ready for a review and I can take another look to see if things are ready for merging. Thanks! -Miles

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,425 @@
+    def sympy(self, index=None):
+        return self._regressor.sympy(index=index)
+
+    def latex(self, index=None, precision=3):
+        return self._regressor.latex(index=index, precision=precision)
```
```suggestion
    def latex(self, **kwargs):
        return self._regressor.latex(**kwargs)
```

Similar suggestion elsewhere. This pattern is good because it means if I update `PySRRegressor.latex` with another parameter, I don't need to also update this one.

MilesCranmer replied:
See https://refactoring.guru/smells/shotgun-surgery – basically want to avoid code patterns that result in needing to do "shotgun surgery"

MilesCranmer commented:
Also I just noticed – `environment.yml` should have the numpy version updated to 1.20 minimum

wenbang24 committed:
removed julia properties

wenbang24 committed:
changed from_file to use PySRRegressor's from_file

wenbang24 committed:
removed uncessary line

wenbang24 committed:
moved args???

python is weird

wenbang24 committed:
a lot of **kwargs

wenbang24 committed:
removed super().__init__()

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wenbang24 committed:
updated numpy to 1.20 in environment.yml

wenbang24 committed:
changed __repr__ to only change first instance

wenbang24 committed:
updated latex_table to use PySRRegressor's latex_table, also added output_variable_names parameter to latex_table

wenbang24 committed:
removed MultiOutputMixin, RegressorMixin

wenbang24 committed:
fixed bug with output_variable_names

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wenbang24 commented:
@MilesCranmer I think things are good to go; can you review please?

wenbang24 committed:
Merge branch 'master' into recurrence

wenbang24 committed:
add back super().__init__()

slightly confused

Co-authored-by: Miles Cranmer <miles.cranmer@gmail.com>

wenbang24 committed:
stars and stuff

Co-authored-by: Miles Cranmer <miles.cranmer@gmail.com>

wenbang24 committed:
another star

Co-authored-by: Miles Cranmer <miles.cranmer@gmail.com>

wenbang24 committed:
remove unused imports

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,299 @@
+        self.n_features = X.shape[1]  # for latex_table()
+
+        current_X = X[self.recursive_history_length :]
+        historical_X = np.lib.stride_tricks.sliding_window_view(
+            X[:-1].flatten(), self.recursive_history_length * X.shape[1]
+        )[:: current_X.shape[1], :]
```
Since this code is similar to the one used in `predict`, could we move it to a separate internal function? That way there is only one appearance of `np.lib.stride_tricks.sliding_window_view` and the name of that internal function would make its behavior clear.

wenbang24 replied:
unfortunately in predict() there is no [:-1] but there is a [:-1] in fit()
I'll add a slide_up_to parameter tho?

wenbang24 replied:
actually predict() has [:: X.shape[1] :] but fit() has [:: current_X.shape[1] :]
I'll probably just move the sliding_window_view to another function without the slice

wenbang24 committed:
refactor: moved np.lib.stride_tricks.sliding_window_view out to another function

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wenbang24 committed:
Merge branch 'master' into recurrence

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,300 @@
+            previous_points = historical_X[-1]
+            # Without this, the model will re-predict the last data point
+            pred_once = self._regressor.predict(X=[previous_points], index=index)
+            previous_points = previous_points[X.shape[1] :]
+            previous_points = np.append(previous_points, pred_once)
+            previous_points = previous_points.flatten()
```
Any simpler way to write this?

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,300 @@
+            previous_points = np.append(previous_points, pred_once)
+            previous_points = previous_points.flatten()
+            for _ in range(extra_predictions):
+                pred_once = self._regressor.predict(X=[previous_points], index=index)
+                previous_points = previous_points[X.shape[1] :]
+                previous_points = np.append(previous_points, pred_once)
+                previous_points = previous_points.flatten()
+                output = np.append(output, pred_once)
```
Can this loop handle the 0th iteration too?

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,300 @@
+            Weight array of the same shape as `X`.
+            Each element is how to weight the mean-square-error loss
+            for that particular element of `X`. Alternatively,
+            if a custom `loss` was set, it will can be used
+            in arbitrary ways.
```
Not sure what this part says

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,300 @@
+        complexity_of_variables : int | float | list[int] | list[float]
+            The complexity of each variable in `X`. If a single value is
+            passed, it will be used for all variables. If a list is passed,
+            its length must be the same as recurrence_history_length.
```
```suggestion
            its length must be the same as `recurrence_history_length`.
```

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,300 @@
+            particular row of `self.equations_`, you may specify the index here.
+            For multiple output equations, you must pass a list of indices
+            in the same order.
+        extra_predictions : int
+            If you want to predict more than one step into the future, specify
+            how many extra predictions you want. For example, if `extra_predictions=2`,
+            the model will predict the next two time points after the last time point
+            in `X`.
```
I think it would be simpler to have this parameter be `num_predictions` and the default be `1`.

wenbang24 committed:
update doctoring

Co-authored-by: Miles Cranmer <miles.cranmer@gmail.com>

wenbang24 commented:
*docstring

wenbang24 committed:
update docstring for weights

wenbang24 committed:
rewrote extra predictions in predict() to use num_predictions

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wenbang24 committed:
fixed predicting and changed up tests

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wenbang24 committed:
removed comments

wenbang24 committed:
updated docstring

wenbang24 committed:
updated docstring to remove y

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,299 @@
+            For multiple output equations, you must pass a list of indices
+            in the same order.
+        num_predictions : int = 1
+            How many predictions to make. If `num_predictions` is less than (n_times - recursive_history_length + 1),
```
```suggestion
            How many predictions to make. If `num_predictions` is less than
            `(n_times - recursive_history_length + 1)`,
```

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,299 @@
+            "PySRRegressor", "PySRSequenceRegressor", 1
+        )
+
+    def get_best(self, **kwargs):
+        return self._regressor.get_best(**kwargs)
+
+    def refresh(self, **kwargs):
+        return self._regressor.refresh(**kwargs)
+
+    def sympy(self, **kwargs):
+        return self._regressor.sympy(**kwargs)
+
+    def latex(self, **kwargs):
+        return self._regressor.latex(**kwargs)
```
```suggestion
    def get_best(self, *args, **kwargs):
        return self._regressor.get_best(*args, **kwargs)

    def refresh(self, *args, **kwargs):
        return self._regressor.refresh(*args, **kwargs)

    def sympy(self, *args, **kwargs):
        return self._regressor.sympy(*args, **kwargs)

    def latex(self, *args, **kwargs):
        return self._regressor.latex(*args, **kwargs)
```

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,299 @@
+
+    def latex_table(
+        self,
+        **kwargs,
```
Need `*args` here as well

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,299 @@
+                return [f"xt_{i}" for i in range(self.recursive_history_length, 0, -1)]
+            else:
+                return [
+                    f"x{i}t_{j}"
```
Or even `x1_t1, x1_t2, ...`. This might mesh well with when the user provides custom features, like `a, b, c` would become `a_t1, a_t2, ..., b_t1, b_t2, ..., c_t1, c_t2, ...`.

wenbang24 replied:
oh yeah the original idea was to use the underscore as a minus, but looking back its not really clear.

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,299 @@
+        Returns
+        -------
+        x_predicted : ndarray of shape (n_samples, n_features)
+            Values predicted by substituting `X` into the fitted sequence symbolic
+            regression model.
```
```suggestion
            Values predicted by substituting `X` into the fitted sequence symbolic
            regression model and rolling it out for `recursive_history_length` steps.
```

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,299 @@
+
+        Returns
+        -------
+        x_predicted : ndarray of shape (n_samples, n_features)
```
```suggestion
        x_predicted : ndarray of shape (num_predictions, n_features)
```

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,299 @@
+            particular row of `self.equations_`, you may specify the index here.
+            For multiple output equations, you must pass a list of indices
+            in the same order.
+        num_predictions : int = 1
```
```suggestion
        num_predictions : int
```

Same format as elsewhere

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,299 @@
+            in the same order.
+        num_predictions : int = 1
+            How many predictions to make. If `num_predictions` is less than (n_times - recursive_history_length + 1),
+            some input data at the end will be ignored.
```
```suggestion
            some input data at the end will be ignored.
            Default is `1`.
```

wenbang24 committed:
Update docstring formatting

Co-authored-by: Miles Cranmer <miles.cranmer@gmail.com>

wenbang24 committed:
added a lot of *args

Co-authored-by: Miles Cranmer <miles.cranmer@gmail.com>

wenbang24 committed:
Update predict docstring to add num_predictions times

Co-authored-by: Miles Cranmer <miles.cranmer@gmail.com>

wenbang24 committed:
fix error in predict docstring

Co-authored-by: Miles Cranmer <miles.cranmer@gmail.com>

wenbang24 committed:
remove default in docstring

Co-authored-by: Miles Cranmer <miles.cranmer@gmail.com>

wenbang24 committed:
add default to predict docstring somewhere else

Co-authored-by: Miles Cranmer <miles.cranmer@gmail.com>

wenbang24 committed:
Merge branch 'master' into recurrence

wenbang24 committed:
added args to latex_table

wenbang24 committed:
changed variable name format

wenbang24 commented:
so latex table outputs `x_{t-0}` now. wdyt?

MilesCranmer commented:
Sorry for so many requests, but do you think you could add an example of using it to the documentation? Currently it is undocumented.

MilesCranmer commented:
> so latex table outputs `x_{t-0}` now. wdyt?

Why only `t-0` instead of (`t-1`, `t-2`, `t-3`, ...)?

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,302 @@
+        Parameters
+        ----------
+        X : ndarray | pandas.DataFrame
+            Sequence of shape (n_times, n_features).
```
```suggestion
            Sequence of shape (n_times, n_features) or (n_times,)
```

MilesCranmer committed:
feat: allow 1D input

MilesCranmer committed:
Update pysr/regressor_sequence.py

MilesCranmer committed:
feat: pretty-print sequence index

MilesCranmer commented:
I pushed a few changes to make the printing look nicer:

```
[ins] In [14]: model.fit(X)
[ Info: Started!
98.5%┣████████████████████████████████████████████████████████████████████████████████████████████████████████▍ ┫ 591/600 [00:02<00:00, 307it/s]Expressions evaluated per second: 7.55e+05. Head worker occupation: 12.8%                                                                       Press 'q' and then <enter> to stop execution early.                                                                                             Hall of Fame:                                                                                                                                   ----------------------------------------------------------------------------------------------------------------------------------------------- Complexity  Loss       Score     Equation                                                                                                       1           6.208e+04  1.594e+01  y = x[t-1]                                                                                                    3           1.000e+00  5.514e+00  y = x[t-1] + x[t-1]                                                                                           5           0.000e+00  7.971e+00  y = (x[t-1] + x[t-2]) + x[t-2]                                                                                -----------------------------------------------------------------------------------------------------------------------------------------------
```

Let me know what you think.

wenbang24 commented:
I think that although one equation would simplify things, there are so many recurrence relations with multiple equations that wouldn't work anymore, so I think we should keep multioutput

wenbang24 commented:
also the pretty printing is indeed pretty pretty; thanks

wenbang24 committed:
sequence example

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wenbang24 committed:
update sequence example

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wenbang24 committed:
added markdown example docs

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

MilesCranmer commented on example_sequence.py:
```diff
@@ -0,0 +1,32 @@
+import numpy as np
+
+X = [[1, 2], [3, 4]]
+for i in range(100):
+    X.append([X[-1][0] + X[-2][0], X[-1][1] / X[-2][1]])
+X = np.array(X)
+
+from pysr import PySRSequenceRegressor
+
+model = PySRSequenceRegressor(
+    recursive_history_length=2,  # How many previous values to use
+    # All other parameters are the same as PySRRegressor
+    model_selection="best",  # Result is mix of simplicity+accuracy
+    niterations=40,
+    binary_operators=["+", "*"],
+    unary_operators=[
+        "cos",
+        "exp",
+        "sin",
+        "inv(x) = 1/x",
+        # ^ Custom operator (julia syntax)
+    ],
+    extra_sympy_mappings={"inv": lambda x: 1 / x},
+    # ^ Define operator for SymPy as well
+    elementwise_loss="loss(x, y) = (x - y)^2",
+    # ^ Custom loss function (julia syntax)
+)
+
+model.fit(X)  # no y needed
+
+print(model)
+print(model.latex())
```
This file isn't needed; just the docs are good

MilesCranmer commented on docs/examples_sequence.md:
```diff
@@ -0,0 +1,53 @@
+# Toy Sequence Examples with Code
```
Can you add it to the existing examples file instead?

MilesCranmer commented on docs/examples_sequence.md:
````diff
@@ -0,0 +1,53 @@
+## Preamble
+```python
+import numpy as np
+from pysr import *
````
```suggestion
from pysr import PySRSequenceRegressor
```

MilesCranmer commented on docs/examples_sequence.md:
````diff
@@ -0,0 +1,53 @@
+```
+
+Note that most of the functionality
+of PySRSequenceRegressor is inherited
+from PySRRegressor.
````
Should link the docs pages using the mkdocs functionality to build references

wenbang24 replied:
which page should it be linked to?

wenbang24 replied:
ok for now i linked it to options.md, wdyt?

wenbang24 committed:
removed example sequence file

MilesCranmer commented on docs/examples_sequence.md:
```diff
@@ -0,0 +1,53 @@
+model = PySRSequenceRegressor(
+    recursive_history_length=2,
+    binary_operators=["+", "*"],
+    extra_sympy_mappings={"inv": lambda x: 1 / x},
```
Mistake?

MilesCranmer commented on docs/examples_sequence.md:
```diff
@@ -0,0 +1,53 @@
+        X[-1][0] + X[-2][0],
+        X[-1][1] - X[-2][1]
+    ])
+X = np.array(X)
```
```suggestion
```

MilesCranmer commented on docs/examples_sequence.md:
```diff
@@ -0,0 +1,53 @@
+X = [1, 1]
+for i in range(20):
+    X.append(X[-1] + X[-2])
+X = np.array(X)
```
```suggestion
```

wenbang24 committed:
removed some miskates in examples sequence

MilesCranmer commented on docs/examples_sequence.md:
```diff
@@ -0,0 +1,52 @@
+
+## 2. Multidimensionality
+
+Here we find a 2D recurrence relation
+with two data points at a time.
```
Would be useful if you say the recurrence relation we are searching for here

wenbang24 commented:
I think stuff breaks without np.array(X)

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,320 @@
+            raise ValueError("num_predictions must be greater than 0.")
+        if num_predictions < len(historical_X):
+            warnings.warn(
+                "The number of predictions is less than the number of historical data points. Some will be ignored."
```
What is being ignored here? Sorry I'm not sure I understand this branch of the code.

wenbang24 replied:
so lets say the user inputted `[1, 2, 3, 4, 5]` and `num_predictions = 1`
then historical_X would be
```
[
[1, 2, 3],
[2, 3, 4],
[3, 4, 5]
]
```
since num_predictions is 1, `[2, 3, 4]` and `[3, 4, 5]` are ignored here and are not used to predict anything

MilesCranmer replied:
Got it! Thanks. In this case I don't think we need the warning. It could also be true that some variables are ignored in the equation itself and we don't warn for that either

wenbang24 committed:
updated example sequence

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,320 @@
+                variable_names = self.variable_names[0] + "_{t-0}"
+            else:
+                variable_names = [
+                    variable_name + "_{t-0}" for variable_name in self.variable_names
+                ]
+        else:
+            if self.n_features == 1:
+                variable_names = "x_{t-0}"
+            else:
+                variable_names = [f"x{i}_{{t-0}}" for i in range(self.n_features)]
+        return self._regressor.latex_table(
+            *args, **kwargs, output_variable_names=variable_names
+        )
```
Why are these `t-0` instead of the actual time?

wenbang24 replied:
🤦 my bad

MilesCranmer commented on pysr/regressor_sequence.py:
```diff
@@ -0,0 +1,320 @@
+        *args,
+        **kwargs,
+    ):
+        """Create a LaTeX/booktabs table for all, or some, of the equations.
+
+        Parameters
+        ----------
+        indices : list[int] | list[list[int]]
+            If you wish to select a particular subset of equations from
+            `self.equations_`, give the row numbers here. By default,
+            all equations will be used. If there are multiple output
+            features, then pass a list of lists.
+        precision : int
+            The number of significant figures shown in the LaTeX
+            representations.
+            Default is `3`.
+        columns : list[str]
+            Which columns to include in the table.
+            Default is `["equation", "complexity", "loss", "score"]`.
+
+        Returns
+        -------
+        latex_table_str : str
+            A string that will render a table in LaTeX of the equations.
+        """
```
```suggestion
```

Instead maybe the docstring can just link to the PySRRegressor one? Then there is no duplication.

wenbang24 committed:
moved sequence examples to examples.md

wenbang24 committed:
updated latex_table to use _t instead of _{t-0}

wenbang24 committed:
updated latex_table docstring to refer to PySRRegressor.latex_table

MilesCranmer commented on docs/examples.md:
```diff
@@ -523,7 +523,56 @@ Note that this expression has a large dynamic range so may be difficult to find.
+### 1. Simple Search
+
+Here's a simple example where we
+find the expression `f(n) = f(n-1) + f(n-2)`.
```
LaTeX works in the docs, you can use `$ ... $`

wenbang24 committed:
updated examples to not have X = np.array(X)

wenbang24 committed:
updated docs to use latex

wenbang24 committed:
removed warning if num_predictions < len(historical_X)

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wenbang24 committed:
updated tests for new variable names

wenbang24 committed:
missed a few

wenbang24 committed:
ok all working now

wenbang24 committed:
whoops forgot to remove commetns

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

wenbang24 commented:
yeah it works mb
but the examples look a bit weird with `X.append(X[-1][0] + X[-2][0])`
should i do that or `X = np.append(X, X[-1] + X[-2])`

wenbang24 commented:
also theres a test failing in PySRRegressor and idk what's going on with it, but I think its something to do with a display_variable_names parameter messing up positional arguments

wenbang24 committed:
fixed a test in TestDimensionalConstraints

wenbang24 commented:
ok i stopped the error; is that the right way to fix it?

wenbang24 committed:
removed unecessary print

wenbang24 commented:
how do we fix the typing issue??
im stumped

wenbang24 committed:
fix typing with a cast

is this cheating lol

wenbang24 commented:
yo all checks passed 🎉

wenbang24 commented:
@MilesCranmer anything else I need to do before merging?

MilesCranmer commented:
Sorry for not reviewing the latest yet, got back from holiday recently and still catching up.

In the meantime you might consider some “dogfooding” (https://en.m.wikipedia.org/wiki/Eating_your_own_dog_food) of PySRSequenceRegressor on some real problems that you have in mind? Just by using it on some real problems you will be able to find ways of improving it. Maybe look up some math sequences or simple physics problems and try it out, and see what breaks and what works well. (This is how I develop PySR in general - I’m both a user and a dev!)

Anyways I hope to get to reviewing soon.

wenbang24 committed:
Merge branch 'master' into recurrence

wenbang24 committed:
Merge branch 'master' into recurrence

# Conflicts:
#	.gitignore
#	docs/src/examples.md
#	environment.yml
#	pysr/__init__.py
#	pysr/sr.py
#	pysr/test/test_main.py
#	requirements.txt

wenbang24 committed:
add homogeneous switch & difference fitting & line of best fit seeding
